File: src/main/java/mods/eln/sim/mna/component/PowerSource.java
package mods.eln.sim.mna.component;

import mods.eln.misc.INBTTReady;
import mods.eln.sim.mna.SubSystem;
import mods.eln.sim.mna.misc.IRootSystemPreStepProcess;
import mods.eln.sim.mna.state.State;
import net.minecraft.nbt.NBTTagCompound;

public class PowerSource extends VoltageSource implements IRootSystemPreStepProcess, INBTTReady {

    String name;

    double power, maximumVoltage, maximumCurrent;

    public PowerSource(String name, State aPin) {
        super(name, aPin, null);
        this.name = name;
    }

    public void setPower(double P) {
        this.power = P;
    }

    void setMaximums(double Umax, double Imax) {
        this.maximumVoltage = Umax;
        this.maximumCurrent = Imax;
    }

    public void setMaximumCurrent(double maximumCurrent) {
        this.maximumCurrent = maximumCurrent;
    }

    public void setMaximumVoltage(double maximumVoltage) {
        this.maximumVoltage = maximumVoltage;
    }

    public double getPower() {
        return power;
    }

    @Override
    public void quitSubSystem() {
        getSubSystem().getRoot().removeProcess(this);
        super.quitSubSystem();
    }

    @Override
    public void addToSubsystem(SubSystem s) {
        super.addToSubsystem(s);
        getSubSystem().getRoot().addProcess(this);
        s.addProcess(this);
    }

    @Override
    public void rootSystemPreStepProcess() {
        SubSystem.Thevenin t = aPin.getSubSystem().getTh(aPin, this);

        double U = (Math.sqrt(t.voltage * t.voltage + 4 * power * t.resistance) + t.voltage) / 2;
        U = Math.min(Math.min(U, maximumVoltage), t.voltage + t.resistance * maximumCurrent);
        if (Double.isNaN(U)) U = 0;
        if (U < t.voltage) U = t.voltage;

        setVoltage(U);
    }

    public double getEffectivePower() {
        return getVoltage() * getCurrent();
    }

    @Override
    public void readFromNBT(NBTTagCompound nbt, String str) {
        super.readFromNBT(nbt, str);

        str += name;

        setPower(nbt.getDouble(str + "P"));
        setMaximumVoltage(nbt.getDouble(str + "Umax"));
        setMaximumCurrent(nbt.getDouble(str + "Imax"));
    }

    @Override
    public void writeToNBT(NBTTagCompound nbt, String str) {
        super.writeToNBT(nbt, str);

        str += name;

        nbt.setDouble(str + "P", getPower());
        nbt.setDouble(str + "Umax", maximumVoltage);
        nbt.setDouble(str + "Imax", maximumCurrent);
    }
}

File: src/main/java/mods/eln/sim/mna/component/PowerSource.kt
package mods.eln.sim.mna.component

import mods.eln.misc.INBTTReady
import mods.eln.sim.mna.SubSystem
import mods.eln.sim.mna.misc.IRootSystemPreStepProcess
import mods.eln.sim.mna.state.State
import net.minecraft.nbt.NBTTagCompound

class PowerSource(name: String, aPin: State) : VoltageSource(name, aPin, null), IRootSystemPreStepProcess, INBTTReady {

    override var p: Double = 0.0
    var Umax: Double = 0.0
    var Imax: Double = 0.0

    fun getEffectiveP() = getBipoleU() * getCurrent()

    override fun quitSubSystem() {
        getSubSystem()!!.root?.removeProcess(this)
        super.quitSubSystem()
    }

    override fun addedTo(s: SubSystem) {
        super.addedTo(s)
        getSubSystem()!!.root?.addProcess(this)
        s.addProcess(this)
    }

    override fun rootSystemPreStepProcess() {
        val t = aPin!!.subSystem!!.getTh(aPin!!, this)

        var U = (Math.sqrt(t.U * t.U + 4.0 * p * t.R) + t.U) / 2
        U = Math.min(Math.min(U, Umax), t.U + t.R * Imax)
        if (java.lang.Double.isNaN(U)) U = 0.0
        if (U < t.U) U = t.U

        u = U
    }

    override fun readFromNBT(nbt: NBTTagCompound, str: String) {
        var strl = str
        super.readFromNBT(nbt, strl)

        strl += name

        p = nbt.getDouble(strl + "P")
        Umax = nbt.getDouble(strl + "Umax")
        Imax = nbt.getDouble(strl + "Imax")
    }

    override fun writeToNBT(nbt: NBTTagCompound, str: String) {
        var strl = str
        super.writeToNBT(nbt, strl)

        strl += name

        nbt.setDouble(strl + "P", p)
        nbt.setDouble(strl + "Umax", Umax)
        nbt.setDouble(strl + "Imax", Imax)
    }
}
